### PR TITLE
Implement `::allocate` and `nil` methods

### PR DESCRIPTION
**Pull request by FnControlOption**
Skipped `NilClass#to_c`, `#to_r`, and `#rationalize` because they require `Complex` and `Rational`

Ported mspec's [`SpecVersion`](https://github.com/ruby/mspec/blob/master/lib/mspec/utils/version.rb)

### CODE REVIEW

File: test/support/version.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  def <=>(other)
+    if other.respond_to? :to_int
+      other = other.to_int
+    else
+      other = SpecVersion.new(other.to_s).to_i
+    end
+
+    self.to_i <=> other
+  end
+end
```
Comment (seven1m) [inline test/support/version.rb:55]: Since this is a direct copy of the file from the mspec library, we will need to make a note in the README about using part of mspec, which is MIT. Up until now, we had re-implemented everything that we needed, not directly looking at mspec code. This is the first file where it was just copied.

Comment (seven1m) [reply test/support/version.rb:55]: Oh, and we should add the copyright to the top of the file just to be clear we didn't write this...

File: include/natalie/string_object.hpp [hunk 2]
```diff
@@ -189,6 +189,6 @@ class StringObject : public Object {
     void raise_encoding_invalid_byte_sequence_error(Env *, size_t) const;
 
     String m_string {};
-    Encoding m_encoding { Encoding::UTF_8 };
+    Encoding m_encoding { Encoding::ASCII_8BIT };
```
Comment (seven1m) [inline include/natalie/string_object.hpp:192]: Can you explain why this change was necessary?

File: include/natalie/nil_object.hpp [hunk 3]
```diff
@@ -22,9 +22,14 @@ class NilObject : public Object {
         return s_instance;
     }
 
+    bool _and(Env *, Value);
+    bool _or(Env *, Value);
```
Comment (seven1m) [inline include/natalie/nil_object.hpp:26]: Our convention is to call these `and_method` and `or_method`

**Comment by seven1m** [comment]
This is great work. The PR itself is kinda all over the place... in the future, it would help with PR review if you could keep it to a single logical change...

The change of default encoding on StringObject has me really scratching my head, so I'm hoping we can put that back to UTF8.

File: include/natalie/string_object.hpp (hunk 2 above)
Comment (FnControlOption) [reply include/natalie/string_object.hpp]: `String::allocate` (and `String::new` without an `encoding:` argument) default to ASCII-8BIT; I've found a better way to do this

**Comment by seven1m** [comment]
I rebased this against master and merged each 'SQUASHME' commit with the commit it accompanied. That was a bit of work, so I'll let you do that next time @FnControlOption. :stuck_out_tongue:
